### PR TITLE
Metadata in WSIReader and MetaTensor in WSIDatasets

### LINKED ISSUE #4956: MetaTensor support for pathology wsi patch datasets: `PatchWSIDatatset`, `SlidingPatchWSIDatatse`, and `MaskedPatchWSIDatatset`

### PR DESCRIPTION
**Pull request by drbeh**
Fixes #4956 and #4674 

### Description
This PR implements micron-per-pixel (mpp) resolution calculation for `WSIReader` and all its backend, and adds `mpp` to the metadata to support the requested metadata from pathology WG.
To be consistent with other image readers, `WSIReader` continue to return (image, metadata) pair instead of a MetaTensor (but it would be nice to reconsider it for all the image readers).
It also updates WSIDatasets to return a MetaTensor and reference the required metadata via image.meta.

### Types of changes

- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/data/wsi_reader.py [hunk 1]
```diff
@@ -322,6 +334,17 @@ def get_file_path(self, wsi) -> str:
         """Return the file path for the WSI object"""
         return self.reader.get_file_path(wsi)
 
+    def get_mpp(self, wsi, level: int) -> Tuple[float, float]:
```
Comment (KumoLiu) [inline monai/data/wsi_reader.py:337]: Maybe this `level` can be optional args, default is None then can use `self.level`.

**Review by KumoLiu — COMMENTED** [review]
Hi @drbeh, in `WSIReader`, I would suggest `level` to be optional optional args in `get_size`, `get_downsample_ratio`, `get_patch` and `get_mpp`.
Below is a simple use case:
```
reader = WSIReader(level=3, ...)
wsi = reader.read(file)

size = reader.get_size(wsi)  # more natural use case
size = reader.get_size(wsi, level = 3)

class WSIReader:
    ... 
    def get_size(self, wsi, level: Optional[int] = None) -> Tuple[int, int]:
            """
            Returns the size (height, width) of the whole slide image at a given level.
            Args:
                wsi: a whole slide image object loaded from a file
                level: the level number where the size is calculated
            """
            if level is None:
                level = self.level

            return self.reader.get_size(wsi, level)
```

**Comment by drbeh** [comment]
Hi @KumoLiu, do you have any other comment? Thanks

File: monai/data/wsi_reader.py [hunk 2]
```diff
@@ -103,6 +104,18 @@ def get_file_path(self, wsi) -> str:
...
+        Args:
+            wsi: a whole slide image object loaded from a file
+            level: the level number where the size is calculated
+
+        """
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
     @abstractmethod
     def get_patch(
         self, wsi, location: Tuple[int, int], size: Tuple[int, int], level: int, dtype: DtypeLike, mode: str
```
Comment (KumoLiu) [inline monai/data/wsi_reader.py:121]: Maybe this `level` args can also be optional?

**Comment by Nic-Ma** [comment]
Hi @drbeh ,

As we suppose to fully support `MetaTensor` in Pathology with this PR, could you please help remove the usage of `FromMetaTensor` in the Pathology tests?
https://github.com/Project-MONAI/MONAI/blob/dev/tests/test_wsireader.py#L23
CC @wyli 

Thanks in advance.

File: monai/data/wsi_reader.py (hunk 2 above)
Comment (drbeh) [reply monai/data/wsi_reader.py:121]: Thanks for your comment but no need to do that here. It is a lower level method that user shouldn't use. I rename them and prepend `_` to denote that they are private methods.

**Comment by drbeh** [comment]
Hi @KumoLiu @Nic-Ma, please let me know if you have any other comment. Otherwise, we are all set to merge this PR. Thanks

**Review by Nic-Ma — APPROVED** [review]

**Comment by Nic-Ma** [comment]
Hi @drbeh ,

I see the CI test failed:
```
2022-09-08T00:54:34.0410022Z [2022-09-08T00:54:09.271Z] ======================================================================
2022-09-08T00:54:34.0410435Z [2022-09-08T00:54:09.271Z] FAIL: test_resolution_mpp_0__home_jenkins_agent_workspace_MONAI_premerge_monai_tests_testing_data_temp_CMU_1_tiff_tiff (tests.test_wsireader.TestCuCIM)
2022-09-08T00:54:34.0410709Z [2022-09-08T00:54:09.271Z] ----------------------------------------------------------------------
2022-09-08T00:54:34.0410918Z [2022-09-08T00:54:09.271Z] Traceback (most recent call last):
2022-09-08T00:54:34.0411265Z [2022-09-08T00:54:09.271Z]   File "/opt/conda/lib/python3.8/site-packages/parameterized/parameterized.py", line 533, in standalone_func
2022-09-08T00:54:34.0411476Z [2022-09-08T00:54:09.271Z]     return func(*(a + p.args), **p.kwargs)
2022-09-08T00:54:34.0411829Z [2022-09-08T00:54:09.271Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_wsireader.py", line 428, in test_resolution_mpp
2022-09-08T00:54:34.0412052Z [2022-09-08T00:54:09.271Z]     self.assertTupleEqual(mpp, expected_mpp)
2022-09-08T00:54:34.0412296Z [2022-09-08T00:54:09.271Z] AssertionError: Tuples differ: (1.0, 1.0) != (1000.0, 1000.0)
2022-09-08T00:54:34.0412425Z [2022-09-08T00:54:09.271Z] 
2022-09-08T00:54:34.0412608Z [2022-09-08T00:54:09.271Z] First differing element 0:
2022-09-08T00:54:34.0412744Z [2022-09-08T00:54:09.271Z] 1.0
2022-09-08T00:54:34.0412884Z [2022-09-08T00:54:09.271Z] 1000.0
2022-09-08T00:54:34.0413011Z [2022-09-08T00:54:09.271Z] 
2022-09-08T00:54:34.0413162Z [2022-09-08T00:54:09.271Z] - (1.0, 1.0)
2022-09-08T00:54:34.0413322Z [2022-09-08T00:54:09.271Z] + (1000.0, 1000.0)
2022-09-08T00:54:34.0413525Z [2022-09-08T00:54:09.271Z] ?   +++     +++
2022-09-08T00:54:34.0413646Z [2022-09-08T00:54:09.271Z] 
```
Could you please help take a look?

Thanks.

File: monai/data/wsi_reader.py [hunk 3]
```diff
@@ -623,34 +716,71 @@ def get_level_count(wsi) -> int:
...
+            factor = 1.0
+        elif unit == unit.INCH:
+            factor = 25400.0
+        else:
+            raise ValueError(f"The resolution unit is not a valid tiff resolution or missing: {unit.name}")
+
+        # Here x and y resolutions are rational numbers so each of them is represented by a tuple.
+        yres = wsi.pages[level].tags["YResolution"].value
+        xres = wsi.pages[level].tags["XResolution"].value
+        return (factor * yres[1] / yres[0], factor * xres[1] / xres[0])
```
Comment (KumoLiu) [inline monai/data/wsi_reader.py:782]: Hi @drbeh, the first number represents the numerator of a fraction and the second is the denominator. It seems like  you've got the numerator and denominator backwards.

Comment (KumoLiu) [reply monai/data/wsi_reader.py:782]: And did you missing to multiply by the `downsample_ratio` here? I have tried three backend on the same image and get three different micro-per-pixel, is there something wrong？

File: monai/data/wsi_reader.py [hunk 4]
```diff
@@ -510,34 +569,68 @@ def get_level_count(wsi) -> int:
...
+            factor = 1000.0
+        elif unit == "micrometer":
+            factor = 1.0
+        elif unit == "inch":
+            factor = 25400.0
+        else:
+            raise ValueError(f"The resolution unit is not a valid tiff resolution: {unit}")
+
+        factor *= wsi.level_downsamples[level]
+        return (factor / float(wsi.properties["tiff.YResolution"]), factor / float(wsi.properties["tiff.XResolution"]))
```
Comment (KumoLiu) [inline monai/data/wsi_reader.py:632]: Perhaps it should be multiplication rather than division here?

File: monai/data/wsi_reader.py [hunk 5]
```diff
@@ -388,34 +424,57 @@ def get_level_count(wsi) -> int:
...
+            wsi: a whole slide image object loaded from a file
+            level: the level number where the size is calculated. If not provided the default level (from `self.level`)
+                will be used.
+
+        """
+        if level is None:
+            level = self.level
+
+        factor = float(wsi.resolutions["level_downsamples"][level])
+        return (wsi.metadata["cucim"]["spacing"][1] * factor, wsi.metadata["cucim"]["spacing"][0] * factor)
```
Comment (KumoLiu) [inline monai/data/wsi_reader.py:476]: I'm not sure whether the spacing here is pointing to level 0 or, but I noticed that he doesn't change with level.

Comment (drbeh) [reply monai/data/wsi_reader.py:476]: Spacing here refers to level 0, and that's why we have to multiply it by downsample.

File: monai/data/wsi_reader.py (hunk 3 above)
Comment (drbeh) [reply monai/data/wsi_reader.py:782]: Hi @KumoLiu, thanks for comments. The calculation of mpp is correct. `XResolution` and `YResolution` have inverse relation to mpp so the fraction looks fine. The reason you are getting different numbers have two reason: 1) the test image is not a fully valid TIFF file (that's why I only check at one level) and 2) there was a bug in cucim that it is resolved in the new version.
If you like you can check with one of the images here: https://drive.google.com/drive/folders/0BzsdkU4jWx9BWk11WEtZZUNFY0U?resourcekey=0-U0E7SyHPJeQd77VAi3z15Q

File: monai/data/wsi_reader.py (hunk 4 above)
Comment (drbeh) [reply monai/data/wsi_reader.py:632]: "YResolution" has inverse relation to mpp.

**Review by KumoLiu — APPROVED** [review]
Thanks for @drbeh's explanation. It looks good to me now, thanks!

**Pull request merged**